fix: variable not found #26144

sqewad opened the pull request:
1. variable not found when accessing a nested attr with empty value

> [!IMPORTANT]
>
> 1. Make sure you have read our [contribution guidelines](https://github.com/langgenius/dify/blob/main/CONTRIBUTING.md)
> 1. Ensure there is an associated issue and you have been assigned to it
> 1. Use the correct syntax to link this PR: `Fixes #<issue number>`.

## Summary

fix: #26144
Distinguish between the case where a variable exists but its value is empty and the case where the variable does not exist.

## Screenshots

| Before | After |
|--------|-------|
| ![Weixin Image_20250924084836_137](https://github.com/user-attachments/assets/ab2aeaad-3817-461a-a2c1-cef066a8cfa1) | <img width="2812" height="1542" alt="Weixin Image_20250924084833_138" src="https://github.com/user-attachments/assets/1f6c13f7-40ac-4a1f-947a-c0709bfdc9ec" /> |

## Checklist

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

Commit:
fix:
1. variable not found when accessing a nested attr with empty value

sqewad committed:
Merge branch 'main' into fix_variable_not_found

QuantumGhost commented on api/core/workflow/entities/variable_pool.py:
```diff
@@ -187,8 +187,8 @@ def _extract_value(self, obj: Any):
     def _get_nested_attribute(self, obj: Mapping[str, Any], attr: str):
```
A minor suggestion: instead of returning `tuple[bool, Any]`, consider changing the signature of this function to `tuple[Any, bool]`, and adding a short description about the second value in the tuple to the docstring.

sqewad replied:
sure I'll do it, thanks for the suggestion

QuantumGhost reviewed (commented):
Hi @sqewad, thank you for the contribution!

The overall changes look good to me. However, I'd suggest adding unit tests for both the `get` and `_get_nested_attribute` methods to cover the specific edge case you fixed. 
This will help prevent regression and ensure the nested attribute handling works correctly in various scenarios.

This would make the fix more robust and maintainable.

Commit:
test:
1. add unit test for variable_pool.py
2. Add docstring for _get_nested_attribute

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

sqewad committed:
Merge branch 'main' into fix_variable_not_found

Commit:
fix:
1. fix import for test_variable_pool.py

QuantumGhost commented on api/core/workflow/entities/variable_pool.py:
```diff
@@ -184,11 +184,22 @@ def _extract_value(self, obj: Any):
+
+        Returns:
+            tuple:
+                - A boolean indicating whether the attribute exists in the object.
```
It seems that the order of elements in returned tuple is not updated.

IMO it's better to return the two values as 
`None, False` and `obj.get(attr), attr in obj`

sqewad replied:
Thanks for pointing it out, I’ve updated the tuple order as suggested.

Commit:
reorder tuple elements and update unit tests

Commit:
return a NoneSegment instead of tuple and update unit test

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

laipz8200 committed:
fix(variable_factory): return Segment if value is already a Segment

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
test: cover falsy variable pool segments
